hCNV implementaion studes (CNVkit)

khaled196 opened the pull request:
FOR CONTRIBUTOR:
* [check] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [check ] - License permits unrestricted use (educational + commercial)
* [ check] - This PR adds a new tool or tool collection
* [ check] - This PR updates an existing tool or tool collection
* [ check] - This PR does something else (explain below)

khaled196 commented:
The tool testing execution succeeded with  planemo t --update_test_data


Shut down
supervisord has terminated
All 2 test(s) executed passed. Therefore, no test data were updated.
CNVkit batch (Test #1): passed
CNVkit batch (Test #2): passed

bgruening commented:
Do you need help @khaled196 ?

khaled196 commented:
Thank you, Björn, 

I had a problem with testing the tools on the GitHub workflows. The reason behind that the results were located in the output file inside the test data. I removed the pull request to test something. 

The last problem is one of the CNVkit sub-tools. It creates a pdf file with the date. The test will always fail even though there is no problem with it.

bgruening commented:
@khaled196 you can use a lot more asserts to test your tools like you can see here: https://docs.galaxyproject.org/en/latest/dev/schema.html#tool-tests-test-output-assert-contents

For example you can check for a similar size: https://docs.galaxyproject.org/en/latest/dev/schema.html#has-size

khaled196 commented:
Hello @bgruening, the tools have successfully passed the tests. Is there anything else I have to do to upload those tools into Galaxy?

bgruening commented on tools/cnvkit_access/CNVkit.xml:
```diff
@@ -0,0 +1,46 @@
+<tool id="CNVkit access" name="CNVkit access" version="0.9.9" python_template_version="3.5">
```
remove python_template_version but include the profile="21.05"

bgruening replied:
use your macro @TOOL_VERSION@ here, please see https://galaxy-iuc-standards.readthedocs.io/en/latest/best_practices/tool_xml.html?highlight=version#tool-versions

khaled196 replied:
All of these points had been addressed in the new pull request

bgruening commented on tools/cnvkit_access/CNVkit.xml:
```diff
@@ -0,0 +1,46 @@
+        ]]></configfile>
+    </configfiles>
+    <inputs>
+        <param name="fa_fname" type="data" format="fasta" multiple="false" label="Target coverage file" help="(.targetcoverage.cnn)." />
```
multiple="false" can be removed, the help could be made more readable for the user

bgruening commented on tools/cnvkit_access/CNVkit.xml:
```diff
@@ -0,0 +1,46 @@
+            --output '$output_setting.output'
+    ]]></command>
+        <configfiles>
+        <configfile name="script_file"><![CDATA[
```
where do you use this file?

khaled196 replied:
I found out this was a bad interpretation from my side so changed the code in the tools for better interpretation.

bgruening commented on tools/cnvkit_access/CNVkit.xml:
```diff
@@ -0,0 +1,46 @@
+    <help><![CDATA[
+    Calculate the sequence-accessible coordinates in chromosomes from the given reference genome, output as a BED file.
+    ]]></help>
+    <citations>
```
could be moved into the shared macro

khaled196 replied:
I have removed the text in the help part as it repeats the same point in the description.  

All of them have the same reference sole reference, so adding the reference to the macros and expanding it later will be the same thing (in my opinion).

bgruening reviewed (commented):
Thanks @khaled196!

You can put everything in one folder called "cnvkit" and reuse the macro in all tools for example.

Do you need help with git? If so let me know and I can help you.
A few comments inline.

khaled196 commented:
The tools are ready to be reviewed again. 

I also have something to add most of the output files are in .cnn, .cnr and .cns format. I have included them under tabular data type/data format as those formats are not included in the galaxy code base.

bgruening commented:
Its even simpler, sorry, that you spent so much time on it.

Have a look here, it really can go everything in one folder.
https://github.com/galaxyproject/tools-iuc/tree/main/tools/mitos

See also the .shed.yml file in this folder. Planemo is smart enough with such a .shed.yml file to decompose all your tools into seperate toolshed repos.

khaled196 commented:
Thank you for your time @bgruening. I will add them as you proposed and reupdated the repository.

khaled196 commented:
Hi @bgruening, I have combined them into one file. Is there anything more I can do?

bgruening commented on tools/cnvkit/access.xml:
```diff
@@ -0,0 +1,44 @@
+<tool id="cnvkit_access" name="CNVkit Access"  version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="21.05">
+    <description>Calculate the sequence-accessible coordinates in chromosomes from the given reference genome, output as a BED file.</description>
```
try to shorten the description a bit, the longer description can go to the help section.

> Calculate the sequence-accessible coordinates in chromosomes

Is probably enough.

bgruening commented on tools/cnvkit/access.xml:
```diff
@@ -0,0 +1,44 @@
+    ]]></command>
+    <inputs>
+        <param name="fa_fname" type="data" format="fasta" label="Target coverage file" help="Example genome.fasta" />
+        <param name="min_gap_size" optional="true" type="integer" label="Minimum gap size between accessible sequence regions." value="5000" help="Regions separated by less than this distance will be joined together.(only use integers, don't use fractures)" />
```
is it really optional? because you set a default value and if here is nothing, then your command probably breaks.

You can also use `argument="--min-gap-size"` here. Instead of using "name". Galaxy will create automatically a nice help addition for you and create the variable `min_gap_size` automatically.

bgruening commented on tools/cnvkit/access.xml:
```diff
@@ -0,0 +1,44 @@
+    <inputs>
+        <param name="fa_fname" type="data" format="fasta" label="Target coverage file" help="Example genome.fasta" />
+        <param name="min_gap_size" optional="true" type="integer" label="Minimum gap size between accessible sequence regions." value="5000" help="Regions separated by less than this distance will be joined together.(only use integers, don't use fractures)" />
+        <param name="exclude" type="data" optional="true" format="bed" multiple="false" label="EXCLUDE" help="Additional regions to exclude, in BED format. Can be used multiple times." />
```
multiple=false is the defaul and can be removed

bgruening replied:
> Can be used multiple times.

Should it be multiple=true?

bgruening commented on tools/cnvkit/access.xml:
```diff
@@ -0,0 +1,44 @@
+            <output name="out_sample_access" file="access-excludes.bed" /> 
+        </test>
+    </tests>
+    <help><![CDATA[
```
Pleaes add a help text for all your tools.

bgruening commented on tools/cnvkit/access.xml:
```diff
@@ -0,0 +1,44 @@
+        <param name="fa_fname" type="data" format="fasta" label="Target coverage file" help="Example genome.fasta" />
+        <param name="min_gap_size" optional="true" type="integer" label="Minimum gap size between accessible sequence regions." value="5000" help="Regions separated by less than this distance will be joined together.(only use integers, don't use fractures)" />
+        <param name="exclude" type="data" optional="true" format="bed" multiple="false" label="EXCLUDE" help="Additional regions to exclude, in BED format. Can be used multiple times." />
+        <section name="output_setting" title="Change output file name" expanded="false">
```
You don't need to do that. The filename is not nessarry to be set by the user, isn't it.

You can just write `--output access-excludes.bed` in the command section.

bgruening commented on tools/cnvkit/access.xml:
```diff
@@ -0,0 +1,44 @@
+    <help><![CDATA[
+    ]]></help>
+    <citations>
+        <citation type="doi">10.1371/journal.pcbi.1004873</citation>
```
can this be moved into the macro file?

bgruening commented on tools/cnvkit/autobin.xml:
```diff
@@ -0,0 +1,87 @@
+                samtools faidx ./genome.fa 2>&1 || echo 'Error running samtools faidx for indexing fasta reference for CNVkit' >&2 &&
+            #else
+                ln -s '$advanced_settings.reference_source.fasta.fields.path' ./genome.fa &&
+                cp '${advanced_settings.reference_source.fasta.fields.path}.fai' ./genome.fa.fai &&
```
you can also link that one, isn't it?

bgruening commented on tools/cnvkit/batch.xml:
```diff
@@ -0,0 +1,249 @@
+                samtools faidx ./genome.fa 2>&1 || echo 'Error running samtools faidx for indexing fasta reference for CNVkit' >&2 &&
+            #else
+                ln -s '$CNV_reference.reference_source.fasta.fields.path' ./genome.fa &&
+                cp '${CNV_reference.reference_source.fasta.fields.path}.fai' ./genome.fa.fai &&
```
link?

bgruening commented on tools/cnvkit/batch.xml:
```diff
@@ -0,0 +1,249 @@
+        </data>
+    </outputs>
+    <tests>
+        <test expect_num_outputs="13">
```
Very good!

bgruening commented on tools/cnvkit/call.xml:
```diff
@@ -0,0 +1,82 @@
+            ./tumor.cns
+            #if $advanced_settings.method == "threshold"
+                #set $method_val = "threshold"
+            --method '$method_val'
```
indentation is a bit wired here

bgruening commented on tools/cnvkit/call.xml:
```diff
@@ -0,0 +1,82 @@
+            --zygosity-freq $additional_SNP_allelic_process.normal_id
+    ]]></command>
+        <configfiles>
+        <configfile name="script_file"><![CDATA[
```
is this file needed?

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,210 @@
+                <option value="wgs"> whole genome sequencing </option>
+            </param>
+            <param name="segment_method" type="select" label="Method used in the 'segment' step" help="">
+                <option value="cbs" selected="True">Circular Binary Segmentation (CBS) method,hybrid (CBS)</option>
```
two time (CBS)? a space is missing

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,210 @@
+            <param name="drop_low_coverage" type="boolean" checked="false" truevalue="TRUE" falsevalue="FALSE" label="Drop very-low-coverage bins before segmentation" help="To avoid false-positive deletions in poor-quality tumor samples" />
+    </xml>
+    <xml name="create_CNV_reference_file">
+        <param name="input_sample_file" type="data" format="bam" label="Sample BAM file" help="Sample file in .BAM format." />
```
I think you can remove the DOT from the formats

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,210 @@
+    </xml>
+    <xml name="output_section">
+        <section name="output_section" title="Outputs" expanded="false">
+            <param name="scatter" type="boolean" checked="false" truevalue="TRUE" falsevalue="FALSE" label="Create a whole-genome copy ratio profile as a PNG scatter plot." help="" />
```
You can use a trick with all booleans.

`<param argument="--scatter" type="boolean" checked="false" truevalue="--scatter" falsevalue="" label="Create a whole-genome copy ratio profile as a PNG scatter plot." help="" />`

And then you can use it in the command section like:

`$output_section.scatter`

Without any if/else or without `--scatter` because the variable is wither `--scatter` in the truecase or `` in the falsecase

bgruening reviewed (commented):
Impressive work @khaled196! Very cool, I added a few comments, mostly with tips and tricks for you to make things shorter or more standard. Those things apply to all tools.

Thanks!

khaled196 commented:
Thank you @bgruening. Kindly find the last amendments to the files.

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+        <param argument="--short-names" type="boolean" checked="false" truevalue="--short-names" falsevalue="" label="Reduce multi-accession bait labels" help="" />
+        <param argument="--target-avg-size" type="integer" optional="true" label="Average size of split target bins" value="0" help="Only integers, don't use fractures" />
+        <param argument="--access" optional="true" type="data" format="bed" label="Regions of accessible sequence on chromosomes (BED)" help="" />
+        <param argument="--antitarget-avg-size" optional="true" type="integer" label="Average size of antitarget bins" value="0" help="" />
```
is there a min/max that makes sense here?

khaled196 replied:
There is no default for those in the documentation. I have changed them from value="0" to value="".

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+    </xml>
+    <xml name="reuse_an_existing_cnv_reference_file">
+        <param name="input_sample_file" type="data" multiple="true" format="bam" label="Sample file" help="Sample BAM ile" />
+        <param argument="--reference" type="data" format="tabular" label="CNV reference CNN File" help="Reference file in CNN format." />
```
Do we have documentation about a CNN file? I think its a tabular, correct? Can we describe which columns it should have? Maybe that can go to the tool help section of a few tools?

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+                <option value="sem"></option>
+            </param>
+            <param argument="--thresholds" optional="true" type="text" label="Hard thresholds for calling each integer copy number, separated by commas." value="=-1.1,-0.25,0.2,0.7" help=" Use the '=' sign on the command line, e.g.: -t=-1,0,1" />
+            <param argument="--ploidy" optional="true" type="integer" label="Ploidy of the sample cells." value="2" help="" />
```
min/max possible?

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+                <option value="ci"></option>
+                <option value="sem"></option>
+            </param>
+            <param argument="--thresholds" optional="true" type="text" label="Hard thresholds for calling each integer copy number, separated by commas." value="=-1.1,-0.25,0.2,0.7" help=" Use the '=' sign on the command line, e.g.: -t=-1,0,1" />
```
the help is not useful in the Galaxy context

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+        <param argument="--sample-id" optional="true" type="text" label="Name of the sample in the VCF to use for b-allele frequency extraction." value="" help="" />
+        <param argument="--normal-id" optional="true" type="text" label="Corresponding normal sample ID in the input VCF" value="" help="This sample is used to select only germline SNVs to calculate b-allele frequencies." />
+        <param argument="--min-variant-depth" optional="true" type="integer" label="Minimum read depth for a SNV to be used in the b-allele frequency calculation." value="20" help="" />
+        <param argument="--zygosity-freq" optional="true" type="float" label="Ignore VCF's genotypes" value="0.25" help="Ignore VCF's genotypes and instead infer zygosity from allele frequencies." />
```
```suggestion
        <param argument="--zygosity-freq"  type="float" value="0.25" optional="true" label="Ignore VCF's genotypes and instead infer zygosity from allele frequencies" help="" />
```

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+    <xml name="diagram_optional">
+            <param argument="--segment" optional="true" type="data" format="cns" label="Segment" help="Segmentation calls (.cns), the output of the 'segment' command." />
+            <param argument="--threshold" optional="true" type="float" label="Threshold" value="0.5" help="Copy number change threshold to label genes." />
+            <param argument="--min-probes" optional="true" type="integer" label="Minimum propes" value="3" help="Minimum number of covered probes to label a gene." />
```
min=0 ?

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+    </xml>
+    <xml name="diagram_optional">
+            <param argument="--segment" optional="true" type="data" format="cns" label="Segment" help="Segmentation calls (.cns), the output of the 'segment' command." />
+            <param argument="--threshold" optional="true" type="float" label="Threshold" value="0.5" help="Copy number change threshold to label genes." />
```
min=0?

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+        <param argument="--zygosity-freq" optional="true" type="float" label="Ignore VCF's genotypes" value="0.25" help="Ignore VCF's genotypes and instead infer zygosity from allele frequencies." />
+    </xml>
+    <xml name="diagram_optional">
+            <param argument="--segment" optional="true" type="data" format="cns" label="Segment" help="Segmentation calls (.cns), the output of the 'segment' command." />
```
Is CNS a kown Galaxy format?

bgruening replied:
If not we can make one. I can help with that.

bgruening replied:
Then we should add them :)
You only need to add something like that here: https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/config/sample/datatypes_conf.xml.sample#L221

Or you just use `tabular` here.

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+                <option value="Female">Female sample's chromosomal sex</option>
+            </param>
+            <param argument="--no-shift-xy" type="boolean" checked="false" truevalue="--no-shift-xy" falsevalue="" label="Don't adjust the X and Y chromosomes according to sample sex." help="" />
+            <param argument="--chromosome" optional="true" type="text" label="CHROMOSOME" value="" help="Chromosome to display, e.g. 'chr1' (no chromosomal range allowed)" />
```
Label is not that useful. 
```suggestion
            <param argument="--chromosome" type="text" value="" optional="true" label="Chromosome to display" help="e.g. 'chr1', no chromosomal range allowed" />
```

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+    </xml>
+    <xml name="reference_optional">
+            <param argument="--cluster" type="boolean" checked="false" truevalue="--cluster" falsevalue="" label="Calculate and store summary stats for clustered subsets of the normal samples with similar coverage profiles" help="" />
+            <param argument="--min-cluster-size" optional="true" type="integer" label="Minimum cluster size to keep in reference profiles." value="4" help="" />
```
min=1?

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+            <param argument="--segment" optional="true" type="data" format="cns" label="Segment" help="Segmentation calls (.cns), the output of the 'segment' command." />
+            <param argument="--chromosome" optional="true" type="text" label="Chromosome range" value="" help="Chromosome or chromosomal range, e.g. 'chr1' or 'chr1:2333000-2444000'" />
+            <param argument="--gene" optional="true" type="text" label="Name of gene or genes (comma-separated) to display" value="" help="" />
+            <param argument="--range-list" optional="true" type="data" format="bed" label="Range list" help=" File listing the chromosomal ranges to display, as BED" />
```
```suggestion
            <param argument="--range-list" optional="true" type="data" format="bed" label="Range list" help="File listing the chromosomal ranges to display, as BED" />
```

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+            <param argument="--method" type="select" label="Segmentation method" help="">
+                <option value="cbs" selected="True">Circular Binary Segmentation (CBS) method,hybrid (CBS)</option>
+                <option value="flasso">Fused lasso, hybrid (flasso)</option>
+                <option value="haar"> a pure-Python implementation of HaarSeg, a wavelet-based method. Very fast and performs reasonably well on small panels, but tends to over-segment large datasets., hybrid (haar)</option>
```
```suggestion
                <option value="haar">A pure-Python implementation of HaarSeg, a wavelet-based method. Very fast and performs reasonably well on small panels, but tends to over-segment large datasets., hybrid (haar)</option>
```

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+                <option value="haar"> a pure-Python implementation of HaarSeg, a wavelet-based method. Very fast and performs reasonably well on small panels, but tends to over-segment large datasets., hybrid (haar)</option>
+                <option value="none">simply calculate the weighted mean log2 value of each chromosome arm. Useful for testing or debugging, or as a baseline for benchmarking other methods., hybrid (none)</option>
+                <option value="hmm">(experimental) – a 3-state Hidden Markov Model suitable for most samples. Faster than CBS, and slower but more accurate than Haar. Requires the Python package pomegranate, as do the next two methods., hybrid (hmm)</option>
+                <option value="hmm-tumor"> (experimental) – a 5-state HMM suitable for finer-grained segmentation of good-quality tumor samples. In particular, this method can detect focal amplifications within a larger-scale, smaller-amplitude copy number gain, or focal deep deletions within a larger-scale hemizygous loss. Training this model takes a bit more CPU time than the simpler hmm method ...
```
```suggestion
                <option value="hmm-tumor">(experimental) – a 5-state HMM suitable for finer-grained segmentation of good-quality tumor samples. In particular, this method can detect focal amplifications within a larger-scale, smaller-amplitude copy number gain, or focal deep deletions within a larger-scale hemizygous loss. Training this model takes a bit more CPU time than the simpler hmm method., hybrid (hmm-tumor)</option>
```

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,215 @@
+            <param argument="--smooth-cbs" type="boolean" checked="false" truevalue="--smooth-cbs" falsevalue="" label="Perform an additional smoothing before CBS segmentations" help="" />
+    </xml>
+    <xml name="citations">
+        <citations>
```
macros?

bgruening commented on tools/cnvkit/reference.xml:
```diff
@@ -0,0 +1,74 @@
+                samtools faidx ./genome.fa 2>&1 || echo 'Error running samtools faidx for indexing fasta reference for CNVkit' >&2 &&
+            #else
+                ln -s '$advanced_settings.reference_source.fasta.fields.path' ./genome.fa &&
+                cp ln -s '${advanced_settings.reference_source.fasta.fields.path}.fai' ./genome.fa.fai &&
```
That looks wrong.
```suggestion
                ln -s '${advanced_settings.reference_source.fasta.fields.path}.fai' ./genome.fa.fai &&
```

bgruening commented on tools/cnvkit/target.xml:
```diff
@@ -0,0 +1,45 @@
+<tool id="cnvkit_target" name="CNVkit Target" version="@TOOL_VERSION@+galaxy@VERSION_SUFFIX@" profile="21.05">
+    <description>Prepare a BED file of baited regions for use with CNVkit.</description>
```
```suggestion
    <description>Prepare a BED file of baited regions for use with CNVkit</description>
```

bgruening commented on tools/cnvkit/target.xml:
```diff
@@ -0,0 +1,45 @@
+    ]]></command>
+    <inputs>
+        <param name="input_file" type="data" format="bed" label="Input BED file" help="" />
+        <param argument="--annotate" optional="true" type="data" format="bed" multiple="false" label="Use gene models from this file to assign names to the target regions" help="Format: UCSC refFlat.txt or ensFlat.txt file (preferred), or BED, interval list, GFF, or similar" />
```
The description allows way more as the format is specifying.
```suggestion
        <param argument="--annotate" optional="true" type="data" format="bed,gff,tabular" multiple="false" label="Use gene models from this file to assign names to the target regions" help="Format: UCSC refFlat.txt or ensFlat.txt file (preferred), or BED, interval list, GFF, or similar" />
```

khaled196 commented:
> Hi @khaled196!
> 
> Thanks a lot. I added a new round to the comments. You can also accept suggestions if you like. In general, a few more tests might help, especially in tools with a lot of branches (conditionals).
> 
> * Is it correct that files tools/cnvkit/test-data/capture.antitarget.bed are really empty?
> * bai files do not need to be included in the test-data, Galaxy will generate them for you

Hi @bgruening 

Thank you for giving me your time in reviewing and sending comments for those tools. 

I have added more tests as well as more conditions to the tools. 

the antitargets files are empty, the creation might come from a problem with the tool itself. as I tried the data that they used to test the tool and return with the same error. 
I have posted an error on the tool GitHub page
https://github.com/etal/cnvkit/issues/806

bgruening reviewed (commented):
We are getting there :)

khaled196 commented on tools/cnvkit/autobin.xml:
```diff
@@ -0,0 +1,113 @@
+            #end while
+            #silent $names.append( $name )
+            ln -s '$bam' '$name'  &&
+        #end for
```
Hi Bjorn, It gives me this error! 

Long-form placeholders - ${}, $(), $[], etc. are not valid inside expressions. Use them in top-level $placeholders only.
Line 6, column 27

Line|Cheetah Code
----|-------------------------------------------------------------
3   |        #set $x=1
4   |        #for $x, $bam in enumerate($bams):
5   |            #set $name_base = re.sub('[^\w\-_\.]', '_', $bam.element_identifier)
6   |            #set $name = ${name_base}_${x}
                               ^
7   |            #silent $names.append( $name )
8   |            ln -s '$bam' '${name}.bam'  &&
9   |            ln -s '${bam.metadata.bam_index}' '${name}.bai'  &&

bgruening replied:
Upps sorry, try `$bam.metadata.bam_index`

khaled196 commented on tools/cnvkit/access.xml:
```diff
@@ -0,0 +1,64 @@
+            #for $exc in $exclude
+                #set $name_base = str(getattr($exc, 'element_identifier', 'sample')).replace( "/", '-' ).replace( "\t", "-" )
+                #set $name = $name_base
+                #while $name in $names:
```
Hi Bjorn,

The loop used if the tool accepts multiple entries,

I am linking each file ($name) from the files list  ---> add the linked file ($name) to a new list --->  call the new list later when I run the code.  

I used the script in the tool below as a hint. 

https://github.com/galaxyproject/tools-iuc/blob/main/tool_collections/kraken/kraken-mpa-report.xml

bgruening commented on tools/cnvkit/access.xml:
```diff
@@ -0,0 +1,61 @@
+                #set $name_base = re.sub('[^\w\-_\.]', '_', $exc.element_identifier)
+                #set $name = $name_base + '_' + str(x)
+                #silent $names.append( $name )
+                ln -s $exc ${name}.bed  &&
```
```suggestion
                ln -s '$exc' ${name}.bed  &&
```

bgruening commented on tools/cnvkit/antitarget.xml:
```diff
@@ -0,0 +1,47 @@
+    <inputs>
+        <param name="targets_file" type="data" format="bed" label="input BED or interval file" help="" />
+        <param argument="--access" type="data" optional="true" format="bed" label=" Regions of accessible sequence on chromosomes" help="" />
+        <param argument="--avg-size" optional="true" type="integer" label=" Average size of antitarget bins." value="150000" help="" />
```
```suggestion
        <param argument="--avg-size" optional="true" type="integer" label="Average size of antitarget bins" value="150000" help="" />
```

bgruening commented on tools/cnvkit/antitarget.xml:
```diff
@@ -0,0 +1,47 @@
+        <param name="targets_file" type="data" format="bed" label="input BED or interval file" help="" />
+        <param argument="--access" type="data" optional="true" format="bed" label=" Regions of accessible sequence on chromosomes" help="" />
+        <param argument="--avg-size" optional="true" type="integer" label=" Average size of antitarget bins." value="150000" help="" />
+        <param argument="--min-size" optional="true" type="integer" label="Minimum size of antitarget bins." value="25000" help="" />
```
```suggestion
        <param argument="--min-size" optional="true" type="integer" label="Minimum size of antitarget bins" value="25000" help="" />
```

bgruening commented on tools/cnvkit/autobin.xml:
```diff
@@ -0,0 +1,125 @@
+            #set $name_base = re.sub('[^\w\-_\.]', '_', $bam.element_identifier)
+            #set $name = $name_base + '_' + str(x)
+            #silent $names.append( $name )
+            ln -s $bam ${name}.bam  &&
```
```suggestion
            ln -s '$bam' ${name}.bam  &&
```

bgruening commented on tools/cnvkit/autobin.xml:
```diff
@@ -0,0 +1,125 @@
+            #set $name = $name_base + '_' + str(x)
+            #silent $names.append( $name )
+            ln -s $bam ${name}.bam  &&
+            ln -s ${bam.metadata.bam_index} ${name}.bai  &&
```
```suggestion
            ln -s '${bam.metadata.bam_index}' ${name}.bai  &&
```

bgruening commented on tools/cnvkit/autobin.xml:
```diff
@@ -0,0 +1,125 @@
+    ]]></command>
+    <inputs>
+        <param name="bams" type="data" multiple="true" format="bam" label="Input Sample BAM file/s" help="" />
+        <param argument="--targets" type="data" format="bed,tabular" label="Input Potentially targeted genomic regions" help=" e.g. all possible exons for the reference genome. Format: BED, interval list, etc." />
```
```suggestion
        <param argument="--targets" type="data" format="bed,tabular" label="Input Potentially targeted genomic regions" help="e.g. all possible exons for the reference genome. Format: BED, interval list, etc." />
```

bgruening commented on tools/cnvkit/batch.xml:
```diff
@@ -0,0 +1,282 @@
+    <expand macro="xrefs"/>
+    <expand macro="requirements"/>
+    <command detect_errors="exit_code"><![CDATA[   
+        ln -s '$CNV_reference.input_sample_file' ./tumor.bam &&
```
```suggestion
        ln -s '$CNV_reference.input_sample_file' ./tumor.bam &&
        ln -s '$CNV_reference.input_sample_file.metadata.bam_index' ./tumor.bam.bai &&
```

bgruening commented on tools/cnvkit/batch.xml:
```diff
@@ -0,0 +1,282 @@
+            ln -s '$CNV_reference.reference' ./reference.cnn &&
+        #end if 
+        #if str($CNV_reference.CNV_reference_availabel) == "no":
+            ln -s '$CNV_reference.normal' ./normal.bam &&
```
```suggestion
            ln -s '$CNV_reference.normal' ./normal.bam &&
            ln -s '$CNV_reference.normal.metadata.bam_index' ./normal.bam.bai &&
```

bgruening commented on tools/cnvkit/batch.xml:
```diff
@@ -0,0 +1,282 @@
+            #end if
+        #end if
+        #import os
+        #set galaxy_slots = os.getenv("GALAXY_SLOTS", 4)
```
```suggestion
```

bgruening commented on tools/cnvkit/batch.xml:
```diff
@@ -0,0 +1,282 @@
+            #end if
+            $output_section.scatter
+            $output_section.diagram
+            --processes $galaxy_slots
```
```suggestion
            --processes \${GALAXY_SLOTS:-4}
```

bgruening commented on tools/cnvkit/batch.xml:
```diff
@@ -0,0 +1,282 @@
+        <data name="out_sample_bintest" format="tabular" label="${tool.name} on ${on_string}: Sample Bintest " from_work_dir="tumor.bintest.cns" />
+        <data name="out_sample_antitargetcoverage" format="tabular" label="${tool.name} on ${on_string}: Sample Antitarget Coverage" from_work_dir="tumor.antitargetcoverage.cnn" />
+        <data name="out_sample_call" format="tabular" label="${tool.name} on ${on_string}: rounded absolute scale copy numbers integer (not log2 ratios) " from_work_dir="tumor.call.cns" />
+        <data name="out_sample_targetcoverage_cnn" format="tabular" label="${tool.name} on ${on_string}: Sample Target Coverage " from_work_dir="tumor.targetcoverage.cnn" />
```
```suggestion
        <data name="out_sample_targetcoverage_cnn" format="tabular" label="${tool.name} on ${on_string}: Sample Target Coverage" from_work_dir="tumor.targetcoverage.cnn" />
```

bgruening commented on tools/cnvkit/heatmap.xml:
```diff
@@ -0,0 +1,62 @@
+            #set $name_base = re.sub('[^\w\-_\.]', '_', $file.element_identifier)
+            #set $name = $name_base + '_' + str(x)
+            #silent $names.append( $name )
+            ln -s $file ${name}.cns  &&
```
```suggestion
            ln -s '$file' ${name}.cns  &&
```

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,216 @@
+        <param argument="--sample-id" optional="true" type="text" label="Name of the sample in the VCF to use for b-allele frequency extraction" value="" help="" />
+        <param argument="--normal-id" optional="true" type="text" label="Corresponding normal sample ID in the input VCF" value="" help="This sample is used to select only germline SNVs to calculate b-allele frequencies" />
+        <param argument="--min-variant-depth" type="integer" value="20" optional="true" label="Minimum read depth for a SNV to be used in the b-allele frequency calculation" help="" />
+        <param argument="--zygosity-freq"  type="float" min="0" value="0.25" optional="true" label="Ignore VCF's genotypes and instead infer zygosity from allele frequencies" help="" />
```
```suggestion
        <param argument="--zygosity-freq" type="float" min="0" value="0.25" optional="true" label="Ignore VCF's genotypes and instead infer zygosity from allele frequencies" help="" />
```

bgruening replied:
```suggestion
        <param argument="--zygosity-freq" type="float" min="0" value="0.25" optional="true" label="Ignore VCF's genotypes and instead infer zygosity from allele frequencies" help="" />
```

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,216 @@
+        <param argument="--fig-size" optional="true" type="float" label="Width and height of the plot in inches" value="" help="Example 6.4 4.8, the space between the two inputs is important"/>
+    </xml>
+    <xml name="segment_optional">
+            <param argument="--dataframe" type="text" optional="true" label="Data frame" value="" help="File name to save the raw R dataframe emitted by CBS or Fused Lasso, example dataframe.r"/>
```
Mh, that is a strange parameter. I would not expose this to the user.

bgruening replied:
ping @khaled196

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,216 @@
+    <xml name="scatter_plot">
+        <param argument="--antitarget-marker" optional="true" type="text" label="Antitarget marker" value="same as targets" help="Plot antitargets using this symbol when plotting in a selected chromosomal region"/>
+        <param argument="--by-bin" type="boolean" checked="false" truevalue="--by-bin" falsevalue="" label="Plot data x-coordinates by bin indices instead of genomic coordinates" help=""/>
+        <param argument="--segment-color" optional="true" type="text" label="Segment color" value="" help=""/>
```
how does such a color looks like?

khaled196 replied:
This was a new update for the tool, I have added this later. changed it to red nothing changed

bgruening commented on tools/cnvkit/reference.xml:
```diff
@@ -0,0 +1,149 @@
+                #set $name_base = re.sub('[^\w\-_\.]', '_', $input.element_identifier)
+                #set $name = $name_base + '_' + str(x)
+                #silent $names.append( $name )
+                ln -s $input ${name}.cnn  &&
```
```suggestion
                ln -s '$input' ${name}.cnn  &&
```

bgruening commented on tools/cnvkit/scatter.xml:
```diff
@@ -0,0 +1,103 @@
+        ln -s '$advanced_settings.segment' ./segment.cns &&
+        #end if
+        #if $advanced_settings.range_list
+        ln -s '$advanced_settings.range_list' ./list.bed &&
```
```suggestion
            ln -s '$advanced_settings.range_list' ./list.bed &&
```

bgruening commented on tools/cnvkit/scatter.xml:
```diff
@@ -0,0 +1,103 @@
+        ln -s '$advanced_settings.range_list' ./list.bed &&
+        #end if
+        #if $advanced_settings.additional_SNP_allelic_process.vcf
+        ln -s '$advanced_settings.additional_SNP_allelic_process.vcf' ./vcf_file.vcf &&
```
```suggestion
            ln -s '$advanced_settings.additional_SNP_allelic_process.vcf' ./vcf_file.vcf &&
```

bgruening commented on tools/cnvkit/segment.xml:
```diff
@@ -0,0 +1,84 @@
+            ln -s '$additional_SNP_allelic_process.vcf' ./vcf_file.vcf &&
+        #end if
+        #import os
+        #set galaxy_slots = os.getenv("GALAXY_SLOTS", 4)
```
```suggestion
```

bgruening commented on tools/cnvkit/segment.xml:
```diff
@@ -0,0 +1,84 @@
+        cnvkit.py segment
+            ./tumor.cnr
+            --output sample.cns
+            --processes $galaxy_slots
```
```suggestion
            --processes \${GALAXY_SLOTS:-4}
```

bgruening commented on tools/cnvkit/segment.xml:
```diff
@@ -0,0 +1,76 @@
+    </outputs>
+    <tests>
+        <test expect_num_outputs="1">
+            <param name="filename" ftype="tabular" value="tumor.cnr" />
```
ping @khaled196

khaled196 replied:
I have no problem doing the test with the optional files, but the vcf files that work are larger to be accepted on the GitHub repo. I tried to trim them but the tool refused them.

bgruening replied:
I see, ok!

bgruening commented on tools/cnvkit/target.xml:
```diff
@@ -0,0 +1,49 @@
+        <param argument="--annotate" optional="true" type="data" format="bed,gff,tabular" label="Use gene models from this file to assign names to the target regions" help="Format: UCSC refFlat.txt or ensFlat.txt file, BED, interval list, GFF, or similar" />
+        <param argument="--short-names" type="boolean" checked="false" truevalue="--short-names" falsevalue="" label="Reduce multi-accession bait labels to be short and consistent" help="" />
+        <param argument="--split" type="boolean" checked="false" truevalue="--split" falsevalue="" label="Split large tiled intervals into smaller, consecutive targets" help="" />
+        <param argument="--avg-size" type="integer" optional="true" label="Average size of split target bins" value="266" help="" />
```
min="1"?

bgruening replied:
Should we add more min/max values?

khaled196 replied:
Hi @bgruening. I wanted to add more min=1, but I found if the user use 0 as value, the argument will be ignored when running the tool. So I left the rest. If adding min value to the integer arguments will add extra guidance for the users then, yes, I think we should add more. I will add them

bgruening replied:
That is a very good point. Glad you found that. This is because `#if $some_integer_or_float:` is in the case of `0` false.
What you need to do here is for all int/float you need to do `#if str($some_integer_or_float):`

khaled196 replied:
Hi @bgruening 

I have updated the tools and added str. 

I have a problem testing the batch.xml file 

The error code is  (cannot find 'metadata' while searching for 'CNV_reference.input_sample_file.metadata.bam_index'\n")

bgruening reviewed (commented):
@khaled196 you can accept my recommendation here o github or if you like provide me commit access to your repo.
Other than that there are only 2 or 3 comments and than we can merge.

bgruening commented on tools/cnvkit/macros.xml:
```diff
@@ -0,0 +1,216 @@
+        <param argument="--trend" type="boolean" checked="false" truevalue="--trend" falsevalue="" label="Draw a smoothed local trendline on the scatter plot" help=""/>
+        <param argument="--y-max" optional="true" type="integer" label="y-axis upper limit" value="" help=""/>
+        <param argument="--y-min" optional="true" type="integer" label="y-axis lower limit" value="" help=""/>
+        <param argument="--fig-size" optional="true" type="float" label="Width and height of the plot in inches" value="" help="Example 6.4 4.8, the space between the two inputs is important"/>
```
ping @khaled196

bgruening commented:
Try without brakes... Sorry mobile.

khaled196 commented:
Hi @bgruening 

Thank you for the replay, and I am sorry to update the review on Friday. 

The break comes from the plane linit step. more precisely after the line below. 

.. INFO: Found all shed metadata fields required for automated repository creation and/or updates.
Error: Process completed with exit code 1.

I will have a look at it over the weekend and update the script on Monday

bgruening commented:
I think the linter is complaining about:

```
Applying linter inputs... WARNING
.. WARNING: Select parameter [center] has option without text
.. WARNING: Select parameter [filter] has option without text
.. WARNING: Select parameter [sample_sex] has option without text
```

khaled196 commented:
Hi @bgruening 

Thank you, I have updated the script to sort out those errors.

bgruening commented on tools/cnvkit/reference.xml:
```diff
@@ -78,6 +92,15 @@
                 <section name="advanced_settings" title="Advanced settings" expanded="false">
                     <expand macro="reference_interface" />
                     <expand macro="reference_optional" />
+                    <conditional name="Sample_sex">
```
Shouldn't this thing also be a macro that you include everywhere?

bgruening commented on tools/cnvkit/reference.xml:
```diff
@@ -133,6 +156,19 @@
                 <assert_contents><has_text text="chromosome"/></assert_contents>
             </output>
         </test>
+        <test expect_num_outputs="1">
```
Thanks for the new test!

bgruening commented on tools/cnvkit/reference.xml:
```diff
@@ -0,0 +1,185 @@
+                    <expand macro="reference_interface" />
+                    <expand macro="reference_optional" /> 
+                    <conditional name="Sample_sex">
+                        <param name="sex" type="select" label="Sample sex availabel" help="">
```
you now removd the "No" option, but I don't get why or if this is not needed why a conditional is needed here?

khaled196 replied:
I gave a thought to this. filter and sample sex has no default option. for example, if there is no sample sex introduced with the inputs the tool predicts the sample sex. So, I didn't add selected = true to the options. what happened is that the tool takes the first option even if it is not determined. I created the condition so that if the user wants to add a filter or know the sample sex, he triggers the conditions and expands the arguments other than that the tool will ignore the arguments.

khaled196 replied:
Hi @bgruening Sorry for that, I have returned the "no" option.

bgruening commented on tools/cnvkit/call.xml:
```diff
@@ -88,7 +87,6 @@
             </conditional>
             <conditional name="add_filter">
                 <param name="Filter" type="select" label="Add filters" help="">
-                    <option value="no" selected="True">Don't add filters to the analysis</option>
```
If you remove this option, the conditional is not needed isn't it? Why have you removed it?
